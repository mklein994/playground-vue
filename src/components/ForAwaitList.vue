<script setup lang="ts">
import { ref, watchEffect } from "vue";

const props = withDefaults(
  defineProps<{
    max?: number;
  }>(),
  {
    max: 25,
  }
);

async function* iteratorGenerator(max: number) {
  let i = 1;
  while (i <= max) {
    yield i++;
  }
}

async function fizzBuzz(max: number) {
  const list = [];
  for await (let i of iteratorGenerator(max)) {
    list.push(
      i % 15 === 0
        ? "FizzBuzz"
        : i % 3 === 0
        ? "Fizz"
        : i % 5 === 0
        ? "Buzz"
        : i
    );
  }
  return list;
}

const max = ref(props.max);

const list = ref<(string | number)[]>([]);

watchEffect(async () => {
  list.value = await fizzBuzz(max.value);
});
</script>

<template>
  <div class="for-await-list">
    <label for="max">Max:</label>
    <input id="max" v-model="max" type="number" name="max" min="1" />

    <ol>
      <li v-for="(value, index) of list" :key="`${value}${index}`">
        <span :class="{ 'is-number': Number.isInteger(value) }">{{
          value
        }}</span>
      </li>
    </ol>
  </div>
</template>

<style scoped>
.is-number {
  color: dodgerblue;
}
</style>
