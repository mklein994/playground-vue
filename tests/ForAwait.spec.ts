import { flushPromises, mount } from "@vue/test-utils";
import { describe, expect, it } from "vitest";
import { defineComponent, h, Suspense } from "vue";

import ForAwaitList from "@/components/ForAwaitList.vue";

describe("ForAwait", async () => {
  const TestComponent = defineComponent(function TestComponent() {
    return () => h(Suspense, () => h(ForAwaitList));
  });
  const wrapper = mount(TestComponent);

  await flushPromises();

  it.each([
    [1, "1"],
    [2, "2"],
    [3, "Fizz"],
    [4, "4"],
    [5, "Buzz"],
    [6, "Fizz"],
    [7, "7"],
    [8, "8"],
    [9, "Fizz"],
    [10, "Buzz"],
    [11, "11"],
    [12, "Fizz"],
    [13, "13"],
    [14, "14"],
    [15, "FizzBuzz"],
    [16, "16"],
  ])("fizzbuzz: %i -> %s", (index, expected) => {
    expect(wrapper.find(`li:nth-child(${index})`).text()).toBe(expected);
  });
});
