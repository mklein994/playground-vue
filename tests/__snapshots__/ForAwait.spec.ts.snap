// Vitest Snapshot v1

exports[`ForAwait > matches snapshot 1`] = `
"<div class=\\"for-await-list\\" data-v-05665efb=\\"\\"><label for=\\"max\\" data-v-05665efb=\\"\\">Max:</label><input id=\\"max\\" type=\\"number\\" name=\\"max\\" min=\\"1\\" data-v-05665efb=\\"\\">
  <ol data-v-05665efb=\\"\\">
    <li data-v-05665efb=\\"\\"><span class=\\"is-number\\" data-v-05665efb=\\"\\">1</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"is-number\\" data-v-05665efb=\\"\\">2</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"\\" data-v-05665efb=\\"\\">Fizz</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"is-number\\" data-v-05665efb=\\"\\">4</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"\\" data-v-05665efb=\\"\\">Buzz</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"\\" data-v-05665efb=\\"\\">Fizz</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"is-number\\" data-v-05665efb=\\"\\">7</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"is-number\\" data-v-05665efb=\\"\\">8</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"\\" data-v-05665efb=\\"\\">Fizz</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"\\" data-v-05665efb=\\"\\">Buzz</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"is-number\\" data-v-05665efb=\\"\\">11</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"\\" data-v-05665efb=\\"\\">Fizz</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"is-number\\" data-v-05665efb=\\"\\">13</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"is-number\\" data-v-05665efb=\\"\\">14</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"\\" data-v-05665efb=\\"\\">FizzBuzz</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"is-number\\" data-v-05665efb=\\"\\">16</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"is-number\\" data-v-05665efb=\\"\\">17</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"\\" data-v-05665efb=\\"\\">Fizz</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"is-number\\" data-v-05665efb=\\"\\">19</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"\\" data-v-05665efb=\\"\\">Buzz</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"\\" data-v-05665efb=\\"\\">Fizz</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"is-number\\" data-v-05665efb=\\"\\">22</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"is-number\\" data-v-05665efb=\\"\\">23</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"\\" data-v-05665efb=\\"\\">Fizz</span></li>
    <li data-v-05665efb=\\"\\"><span class=\\"\\" data-v-05665efb=\\"\\">Buzz</span></li>
  </ol>
</div>"
`;
